File: prelude/toolchains/android/src/com/facebook/buck/step/StepExecutionResult.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under both the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree and the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree.
 */

package com.facebook.buck.step;

import com.facebook.buck.core.util.immutables.BuckStyleValueWithBuilder;
import com.facebook.buck.util.ProcessExecutor;
import com.google.common.collect.ImmutableList;
import java.util.Optional;

/** Exit code, command and stderr info from the executed step */
@BuckStyleValueWithBuilder
public interface StepExecutionResult {

  int getExitCode();

  ImmutableList<String> getExecutedCommand();

  Optional<String> getStderr();

  Optional<Exception> getCause();

  default boolean isSuccess() {
    return getExitCode() == StepExecutionResults.SUCCESS_EXIT_CODE;
  }

  /** Creates {@code StepExecutionResult} from {@code exitCode} */
  static StepExecutionResult of(int exitCode) {
    return new Builder().setExitCode(exitCode).build();
  }

  /** Creates {@code StepExecutionResult} from {@code exception} */
  static StepExecutionResult of(Throwable exception) {
    return StepExecutionResult.builder()
        .setExitCode(StepExecutionResults.ERROR_EXIT_CODE)
        .setCause((Exception) exception.getCause())
        .build();
  }

  /** Creates {@code StepExecutionResult} from {@code ProcessExecutor.Result} */
  static StepExecutionResult of(ProcessExecutor.Result result) {
    return new Builder()
        .setExitCode(result.getExitCode())
        .setExecutedCommand(result.getCommand())
        .setStderr(result.getStderr())
        .build();
  }

  static Builder builder() {
    return new Builder();
  }

  class Builder extends ImmutableStepExecutionResult.Builder {}

  public default String getErrorMessage() {
    return getCause()
        .map(Throwable::getMessage)
        .or(this::getStderr)
        .orElse(String.format("<failed with exit code %s>", getExitCode()));
  }
}

File: prelude/toolchains/android/src/com/facebook/buck/step/StepExecutionResult.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is dual-licensed under either the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree or the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree. You may select, at your option, one of the
 * above-listed licenses.
 */

package com.facebook.buck.step

import java.util.Optional

/** Exit code, command and stderr info from the executed step */
data class StepExecutionResult(
    val exitCode: Int,
    val stderr: Optional<String>,
    val cause: Optional<Exception>
) {
  constructor(
      exitCode: Int,
      stderr: Optional<String>
  ) : this(exitCode, stderr, Optional.empty<Exception>())

  val isSuccess: Boolean
    get() = exitCode == StepExecutionResults.SUCCESS_EXIT_CODE

  val errorMessage: String
    get() {
      return cause
          .map<String>(Throwable::message)
          .or(this::stderr)
          .orElse(String.format("<failed with exit code %s>", exitCode))
    }

  companion object {
    /** Creates `StepExecutionResult` from `exitCode` */
    @JvmStatic
    fun of(exitCode: Int): StepExecutionResult {
      return StepExecutionResult(
          exitCode,
          if (exitCode == 0) Optional.empty() else Optional.of("Failed to execute isolated step."),
          Optional.empty())
    }

    /** Creates `StepExecutionResult` from `exception` */
    @JvmStatic
    fun of(exception: Throwable): StepExecutionResult {
      return StepExecutionResult(
          StepExecutionResults.ERROR_EXIT_CODE,
          Optional.of("Failed to execute isolated step."),
          Optional.ofNullable(exception.cause as Exception?))
    }
  }
}
